### ISSUE TITLE
Do not save the weights of the model to hparams.yaml

### ISSUE THREAD

**Issue by martinwhl** [Lightning-AI/lightning#11494]
## 🚀 Feature

Do not save the weights of the model (passed as an argument to the `__init__()` function of a `LightningModule` to the `hparams.yaml` file, as saving them to the model checkpoint file should be enough.

### Motivation

For an encoder-decoder style `LightningModule` that looks like this:

```Python
class SomeCustomizedTaskOrFramework(pl.LightningModule):
    def __init__(
        self,
        model: nn.Module,
        loss="mse",
        learning_rate: float = 1e-3,
        **kwargs
    ):
        super(SomeCustomizedTaskOrFramework, self).__init__()
        self.save_hyperparameters()
        self.model = model
        self.task_head = ...
        ...
```

The weights of `model`, passed as an argument of `__init__()`, is saved to both the `hparams.yaml` and the model checkpoint. However, it might not be necessary to save the weights to the yaml file, and this process is very time-consuming when the model is large - it might take more than ten minutes before the first epoch actually begins. I tried `save_hyperparameters(ignore=["model"])`, but I cannot successfully load my checkpoint this way.

cc @borda

**Comment by awaelchli** [comment]
Hello

`save_hyperparameters(ignore=["model"])` is the recommended way. It is also expected that you can't load your LightningModule anymore because of the missing "parameter", but model here should not be considered a hyperparameter and therefore it is correct to exclude it. When loading the model, provide it as input like so:

```py
my_backbone = MyNet()
lightning_module = SomeCustomizedTaskOrFramework.load_from_checkpoint(path, model=my_backbone)
```

Does that make sense? 
Docs: https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#excluding-hyperparameters

**Comment by awaelchli** [comment]
I recommend not going ahead with this feature. Dropping the "model" parameter here automatically will just lead to more confusion when attempting to load a LightningModule later on. Right now, the `save_hyperparameter()` and automatic instantiation with `LM.load_from_checkpoint` is well-defined and flexible enough.

**Comment by martinwhl** [comment]
> Hello
> 
> `save_hyperparameters(ignore=["model"])` is the recommended way. It is also expected that you can't load your LightningModule anymore because of the missing "parameter", but model here should not be considered a hyperparameter and therefore it is correct to exclude it. When loading the model, provide it as input like so:
> 
> ```python
> my_backbone = MyNet()
> lightning_module = SomeCustomizedTaskOrFramework.load_from_checkpoint(path, model=my_backbone)
> ```
> 
> Does that make sense? Docs: [pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#excluding-hyperparameters](https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#excluding-hyperparameters)

I did try `save_hyperparameters(ignore=["model"])`. After that, when I need to load from the saved checkpoint, Python throws a `TypeError` saying that the `model` argument is not passed to `__init__()`.

**Comment by martinwhl** [comment]
> I recommend not going ahead with this feature. Dropping the "model" parameter here automatically will just lead to more confusion when attempting to load a LightningModule later on. Right now, the `save_hyperparameter()` and automatic instantiation with `LM.load_from_checkpoint` is well-defined and flexible enough.

`load_from_checkpoint()` uses the saved checkpoint file instead of the `hparams.yaml`. Saving model to the checkpoint is totally fine; the thing is, I don't think it is necessary to save the weights of the model to a yaml file, which takes quite a long time before the epochs actually begin.

**Comment by martinwhl** [comment]
> Hello
> 
> `save_hyperparameters(ignore=["model"])` is the recommended way. It is also expected that you can't load your LightningModule anymore because of the missing "parameter", but model here should not be considered a hyperparameter and therefore it is correct to exclude it. When loading the model, provide it as input like so:
> 
> ```python
> my_backbone = MyNet()
> lightning_module = SomeCustomizedTaskOrFramework.load_from_checkpoint(path, model=my_backbone)
> ```
> 
> Does that make sense? Docs: [pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#excluding-hyperparameters](https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#excluding-hyperparameters)

Thanks for your suggestion, this should work when the whole pipeline is done in one Python script. However, I might need to load the weights of the model some other time, so I suppose I need to manually call `torch.save()` and `torch.load()`.

**Comment by awaelchli** [comment]
> Thanks for your suggestion, this should work when the whole pipeline is done in one Python script. However, I might need to load the weights of the model some other time, so I suppose I need to manually call torch.save() and torch.load().

Not necessarily. In your production use case, you could define a simple wrapper like this:

```py
class ProductionModelWrapper(nn.Module): 
    def __init__(self):
        # the same name and type as in LightningModule used to train it
        # so that when we load the state dict, all the parameter names will match what they were in LightningModule
        self.model = MyNet()  
    def forward(self, *args, **kwargs):
        return self.model(*args, **kwargs)
```

Now, let's say you trained your model with Lightning and it saved a checkpoint. It can be loaded outside of a Lightning environment in your production Python script: 

```py
prod_model = ProductionModelWrapper()
lightning_checkpoint = torch.load("path/to/lightning/checkpoint.ckpt")
prod_model.load_state_dict(lightning_checkpoint["state_dict"])
```

What do you think of this?

**Comment by martinwhl** [comment]
> > Thanks for your suggestion, this should work when the whole pipeline is done in one Python script. However, I might need to load the weights of the model some other time, so I suppose I need to manually call torch.save() and torch.load().
> 
> Not necessarily. In your production use case, you could define a simple wrapper like this:
> 
> ```python
> class ProductionModelWrapper(nn.Module): 
>     def __init__(self):
>         # the same name and type as in LightningModule used to train it
>         # so that when we load the state dict, all the parameter names will match what they were in LightningModule
>         self.model = MyNet()  
>     def forward(self, *args, **kwargs):
>         return self.model(*args, **kwargs)
> ```
> 
> Now, let's say you trained your model with Lightning and it saved a checkpoint. It can be loaded outside of a Lightning environment in your production Python script:
> 
> ```python
> prod_model = ProductionModelWrapper()
> lightning_checkpoint = torch.load("path/to/lightning/checkpoint.ckpt")
> prod_model.load_state_dict(lightning_checkpoint["state_dict"])
> ```
> 
> What do you think of this?

Yes, I understand. The issue I mainly want to address is stated [here](https://github.com/martinwhl/Informer-PyTorch-Lightning/issues/1).

**Comment by martinwhl** [comment]
Got this warning in version 1.6, guess this issue is already considered:
```
UserWarning: Attribute 'model' is an instance of `nn.Module` and is already saved during checkpointing. It is recommended to ignore them using `self.save_hyperparameters(ignore=['model'])`.
```

**Comment by adeandrade** [comment]
I ended suppressing this new warning and ignoring the recommendation. My work is in multi-task learning and my main module loads 5-10 different modules. Loading them each manually and passing them to `load_from_checkpoint` is quite tedious.

Many times the modules might be different classes so to follow the recommendation I would have to store which class was used for each parameter and store that along with the checkpoint.

The current recommendation seems to work against composability. Why don't we have`hparams.yaml` ignore the module parameters but still store the related information about it?

**Comment by hakobtt** [comment]
Hi,
Actually, the `load_from_checkpoint` method doesn't care about `hparams.yaml` file at all and it loads everything from only the checkpoint file.  `hparams.yaml` is for the logger. When you call `self.save_hyperparameters(ignore=[])` it actually saves the hyperparameters in `hparams` attribute and also writes it in `hparams.yaml` file. If you have some parameters in the ignore list they won't appear in the `hparams` attribute and hence won't be in saved  `hparams.yaml` and in checkpoints as well.
1. If you don't need `hparams.yaml` file at all then you can just do 

```self.save_hyperparameters(logger=False)```

in this case all your parameters will be saved in the `hparams` atribute but  it won't create `hparams.yaml` file so it will work fast even if you have large modules as parameters. 

2. If you do need the `hparams.yaml` for some parameters but you don't want your large modules to be saved there. For this case, there is a workaround of calling the` self.save_hyperparameters()` method twice. 
```
self.save_hyperparameters(ignore=["parameters you want to ignore"], logger=True)
self.save_hyperparameters(ignore=[], logger=False)
```
With the first call you save **only necessary** parameters for your logger in the `hparams` attribute and write in the `hparams.yaml` file. 
With the second call,  you save **all** hyperparameters in the `hparams` attribute but you don't create `hparams.yaml` file with those. Later when checkpoints will be saved `hparams` attribute will have all parameters in it so they will be saved in the checkpoints and will be loaded with `load_from_checkpoint` without extra parameters.

**Comment by davidegraff** [comment]
Can we revive this? I'm encountering a similar issue here as well, and even the lightning docs present a likely case:
```python
class Encoder(nn.Module):
    def __init__(*args, **kwargs):
        ...

class Decoder(nn.Module):
    def __init__(*args, **kwargs):
        ...

class AutoEncoder(pl.LightningModule):
    def __init__(self, encoder: Optional[Encoder] = None, decoder: Optional[Decoder] = None):
        self.encoder = encoder or Encoder()
        self.decoder = decoder or Decoder()
```

It's natural to express an `AutoEncoder` as a system composed of an `Encoder`/`Decoder` pair. Beyond the hyperparameters of the broad system (learning rate, KL term weight, etc...), we would also consider the underlying _architectures_ of our encoder/decoder pair as a hyperparameter in the system whereas the _weights_ of the pair would be our parameters. It's reasonable that a user would like to have these all logged in TB or WandB, but under the current system, it's impossible without moving the modules _inside_ of our `AutoEncoder` system, like so:
```python
class AutoEncoder(pl.LightningModule):
    def __init__(self, enc_config: Optional[Mapping] = None, dec_config: Optional[Mapping] = None):
        self.encoder = Encoder(**(enc_config or {}))
        self.decoder = Decoder(**(enc_config or {}))
```
This works, but it's fairly brittle and against the spirit of composability.

Suppose I wanted to adopt the same broad system to use a `CustomEncoder`, I have three choices (1) deal with the fact that I can't easily record the hyperparameters in my logging system; (2) use a workaround like `self.logger.log_hparams`, which would logically correspond to the encoder/decoder architectures being properties of my logger _rather_ than my model (and would also not be able to be called in `__init__()`), or (3) restrict my encoder/decoder architectures and move their arguments to _inside_ the `AutoEncoder`. None of these seems like a good option.

Ideally, I can manually store some hyperparameters of the models in the `save_hyperparameters` function or (better) have them stored for me (provided the input `Module`s implement some `"Detectable"` interface)

**Comment by mattcleigh** [comment]
Also in favour of reviving this, running into the exact same issue as @davidegraff trying to compose and reload many submodules.

**Comment by YaLTeR** [comment]
I'm also facing this problem. I went with passing submodule types and args as strings into the LightningModule and instantiating then in `__init__` approach. This works reasonably well, I get sensible hyperparameters in the logger (i.e. encoder type as string) without having to manually save anything.

However, the issue I'm facing is that I'm passing some of the data as file paths to .npy for the model to load tensors from. These end up stored in the state dict since I register them as nn.Module buffers, but they also end up stored as file paths in hyper parameters. So when I try to load the checkpoint for inference on a different machine, it tries to load those files from their paths (which don't exist on the other machine), despite them already being contained in the state dict. I can't simply ignore those arguments out of hyper parameters either because they are inside argument dicts for submodules (and because this would mean hardcoding submodule-specific argument names in the LightningModule).

Overall I'm not sure if there's any way to solve this cleanly with the current pytorch-lightning facilities.

**Comment by Borda** [comment]
in your model, you shall be able to call save hyperparameters and list args that shall be saved

**Comment by YaLTeR** [comment]
I have `def __init__(self, encoder_type: str, encoder_args: dict)`, and the problematic file path args are in the `encoder_args`, but only for some Encoder types and not the others.
